File: BinarySearch2DMatrix.java
class BinarySearch2DMatrix {
	/*
	Time Complexity: O(n)
	Space Complexity: O(1)
	 */
	public boolean searchMatrix(int[][] matrix, int target) {
		//initialize the variables
		int i=0,j=0;

		//iterate only the outer loop.
		//Check for the last value from the matrix column and look into the row only if last val > target
		for(i = 0; i < matrix.length; i++) {

			//check if target < last element of a row
			if (target <= matrix[i][matrix[0].length-1]) {
				//if true -> iterate/traverse the row else do not traverse the row.
				for (j=0; j< matrix[0].length;j++){
					if (target == matrix[i][j]){
						return true;
					}
				}
			}
		}
		return false;
	}
}

File: binarySearch_2Dmatrix.py
class binarySearch:
     
    def searchMatrix(self, matrix, target):
        r = len(matrix)
        c = len(matrix[0])

        l, h = 0, (r * c) - 1
        
        while l <= h:
            mid = l + (h - l) // 2

            row = mid // c
            col = mid % c

            if (matrix[row][col] == target):
                return True

            elif(matrix[row][col] > target):
                h = mid - 1

            else:
                l = mid + 1

        return False


inputArr = [[1,3,5,7],[10,11,16,20],[23,30,34,60]]
target = 3
input = binarySearch()
res = input.searchMatrix(inputArr, target)
print(res)